### PR TITLE
stdlib audit: atomic refcounting for Bigarray proxy

### PR DESCRIPTION
**Pull request by Octachron**
This PR implements naive atomic refcounting for bigarray proxy.
The atomic refcounting functions are factorized between `io.c` and `bigarray.c` in order to facilitate the switch to a better implementation.

The second commit contains a test that fails currently on trunk but which is too slow to keep beyond the lifetime of this PR.
(See #10832 for more information.)

### CODE REVIEW

File: runtime/bigarray.c [hunk 1]
```diff
@@ -154,7 +155,7 @@ CAMLexport void caml_ba_finalize(value v)
     if (b->proxy == NULL) {
       free(b->data);
     } else {
-      if (-- b->proxy->refcount == 0) {
+      if (caml_atomic_refcount_decr(&b->proxy->refcount) == 0) {
```
Comment (xavierleroy) [inline runtime/bigarray.c]: I thought `refcount_decr` returns the value before decrement.  In this case it should be `== 1`, not `== 0`.

Comment (Octachron) [reply runtime/bigarray.c]: Indeed, that was a stupid mistake. Fixed.

**Review by xavierleroy — COMMENTED** [review]
This looks good to me w.r.t. bigarray handling.

For channels, I'm not convinced the refcount should be atomic, once proper locking is implemented (see #11171).

In particular, `caml_atomic_refcount_get` is a code smell: it makes no sense to check the value of a reference count if it can be modified concurrently; the only case it makes sense is if locking prevents changes, in which case it doesn't need to be atomic.

**Comment by Octachron** [comment]
I propose to remove the channel change for this PR for now, in order to move forward with the bigarray fix.

**Review by xavierleroy — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**